File: 프로그래머스_수식최대화.cpp
#include <string>
#include <vector>
#include <iostream>
#include <cmath>
using namespace std;

long long solution(string expression) {
    char cased[6][3]={{'+','-','*'},
                     {'+','*','-'},
                     {'-','*','+'},
                     {'-','+','*'},
                     {'*','+','-'},
                     {'*','-','+'},};
    vector <string> oper;
    vector <long long> numbers;
    string num="";
    long long Max=-1;
    for(int i=0;i<expression.length();i++){
        if(expression[i]=='+'||expression[i]=='*'||expression[i]=='-'){
            string temp;
            temp.push_back(expression[i]);
            oper.push_back(temp);
            if(num.length()!=0){
                numbers.push_back((long long )stoi(num));
                num="";
            }
        }
        else
            num+=expression[i];
    }
    numbers.push_back((long long)stoi(num));
   for(int i=0;i<6;i++){
        vector <string> tmp_oper=oper;
        vector <long long > tmp_numbers=numbers;
       for(int j=0;j<3;j++){
           for(int l=0;l<tmp_oper.size();l++){
                if(cased[i][j]==tmp_oper[l][0]){
                    if(tmp_oper[l][0]=='+')
                        tmp_numbers[l]=tmp_numbers[l]+tmp_numbers[l+1];
                    else if(tmp_oper[l][0]=='*'){
                        tmp_numbers[l]=tmp_numbers[l]*tmp_numbers[l+1];}
                    else if(tmp_oper[l][0]=='-'){
                        tmp_numbers[l]=tmp_numbers[l]-tmp_numbers[l+1];}
                    tmp_oper.erase(tmp_oper.begin()+l);
                    tmp_numbers.erase(tmp_numbers.begin()+l+1);
                    l--;
               }
           }
       }
       Max=Max>abs(tmp_numbers[0])?Max:abs(tmp_numbers[0]);
   }
    return Max;
}

File: programmers/프로그래머스_수식최대화.py
def solution(expression):
    Max=0
    case= [['+','-','*'],
           ['+','*','-'],
           ['-','+','*'],
           ['-','*','+'],
           ['*','-','+'],
           ['*','+','-'],
          ]
    a=expression.replace("+"," ").replace("-"," ").replace('*',' ')
    for i in case:
        oper=[i for i in expression if i=="+" or i=='-' or i=='*']
        num=list(map(int,a.split(" ")))
        for op in i:
            j=0
            while j<len(oper):
                if op==oper[j]:
                    if oper[j]=='+':
                        num[j]=num[j]+num[j+1]
                    elif oper[j]=='-':
                        num[j]=num[j]-num[j+1]
                    elif oper[j]=='*':
                        num[j]=num[j]*num[j+1]
                    del num[j+1]
                    del oper[j]
                    j-=1
                j+=1
        Max=max(abs(num[0]),Max)
    return Max
